csrc/fusion_segmenter.cpp

```diff
@@ -4346,63 +4346,223 @@ void SegmentCandidateFinder::findSegments() {
+    Expr* old_expr,
+    Expr* new_expr) {
+  // Check if old_expr is a key
+  auto it = privatized_ops.find(old_expr);
+  if (it != privatized_ops.end()) {
+    // Copy the set of values
+    auto cloned_set = it->second;
+    // Remove old entry
+    privatized_ops.erase(it);
+    // Insert with new key
+    privatized_ops[new_expr] = cloned_set;
+    return;
+  }
 
+  // Check if old_expr is in any value sets and update it
+  for (auto& kv : privatized_ops) {
+    if (kv.second.count(old_expr) > 0) {
+      kv.second.erase(old_expr);
+      kv.second.insert(new_expr);
+      return;
+    }
+  }
+}
+
+std::vector<Expr*> get_upcasts_and_squeezes(SegmentedGroup* group) {
+  std::vector<Expr*> upcasts_or_squeezes;
+  std::copy_if(
+      group->exprs().begin(),
+      group->exprs().end(),
+      std::back_inserter(upcasts_or_squeezes),
+      [](Expr* expr) {
+        return ir_utils::isTvOp(expr) &&
+            (is_upcast_op(expr) || expr->isA<SqueezeOp>());
+      });
+
+  return upcasts_or_squeezes;
+}
+
+//!
+//!  Determines if any expanded dimensions in a TensorView need to be squeezed.
+//!
+//!  This function checks each dimension of the provided TensorView `x`
+//!  (excluding reductions) and determines if any dimension marked as `true` in
+//!  the `to_squeeze` vector is both expanded and needs to be squeezed. If at
+//!  least one such dimension exists, the function returns true; otherwise, it
+//!  returns false.
+//!
+//!  @param x Pointer to the TensorView whose dimensions are to be checked.
+//!  @param to_squeeze A boolean vector indicating which dimensions should be
+//!  considered for squeezing.
+//!                    Each element corresponds to a dimension in `x` (after
+//!                    reductions are removed). If an element is true, that
+//!                    dimension is a candidate for squeezing.
+//!  @return true if at least one expanded dimension needs to be squeezed, false
+//!  otherwise.
+//!
+bool needs_to_squeeze_expanded(
+    TensorView* x,
+    const std::vector<bool>& to_squeeze) {
+  auto x_dom = x->domain()->noReductions();
+  const auto ndims = static_cast<int64_t>(x_dom.size());
+  for (const auto idx : arange(ndims)) {
+    // If the dimension is not expanded, no need to squeeze
+    if (!to_squeeze[idx]) {
+      continue;
+    }
+
+    // If the dimension is expanded, we need to squeeze it
+    if (x_dom[idx]->hasExpandedExtent()) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+} // namespace
+
+// In addition to privatizing upcasts, this function also privatizes squeeze.
+// We decided to privatize squeeze ops motivated by cross-entropy loss.
+// Let's say we have the sequence:
+// T1 (input) bfloat -> (up)cast to float  ->  squeeze.
+// also, say, T1 is input to another operation say takeAlongAxis.
+// Since T1 has multiple uses, the T1->cast->squeeze cannot be forwarded
+// and the cast cannot be privatized without privatizing the squeeze.
```

Comment:
I'm not following this. I see that this patten won't enable the input forwarding, but why privatizing the cast is important? Why can't it be done without privatizing squeeze?

Reply:
1. Without privatizing cast the data written out of segment 1 (made up) is float, and this is used in other segments. With privatizing the intermediate become bf16 reducing memory traffic.

In this case, we have something like:
In -> cast -> In0 -> Squeeze -> In1 -> Use1
                             -> In2 -> Use2

Since there is only one use of the output of cast, it is not a candidate of privatization till squeeze is privatized.